tests/credentials_provider_sso_tests.c

```diff
@@ -0,0 +1,443 @@
+    struct aws_string *actual_home = aws_string_new_from_c_str(allocator, getenv("HOME"));
+    ASSERT_SUCCESS(aws_set_environment_value(s_home_env_var, s_home_here));
+
+    /* Remove stale cache directory. */
+    struct aws_string *fake_cache_dir = aws_string_new_from_c_str(allocator, ".aws/sso/cache");
+    ASSERT_SUCCESS(aws_directory_delete(fake_cache_dir, /* recursive = */ true));
+    aws_string_destroy(fake_cache_dir);
+
+    ASSERT_SUCCESS(s_sso_provider_init_config_profile(allocator, sso_test_profile_contents));
+
+    ASSERT_PTR_EQUALS(NULL, aws_credentials_provider_new_sso(allocator, &options));
+
+    /* Clean up */
+    ASSERT_SUCCESS(aws_set_environment_value(s_home_env_var, actual_home));
+    aws_string_destroy(actual_home);
+
+    aws_auth_library_clean_up();
+
+    return AWS_OP_SUCCESS;
+}
+AWS_TEST_CASE(
+    credentials_provider_sso_provider_access_token_file_missing,
+    s_credentials_provider_sso_provider_access_token_file_missing);
+
+static int s_credentials_provider_sso_provider_access_token_expired(struct aws_allocator *allocator, void *ctx) {
+    (void)ctx;
+    struct aws_credentials_provider_profile_options options = {0};
+    AWS_STATIC_STRING_FROM_LITERAL(
+        sso_expired_access_token,
+        "{\
+          \"startUrl\": \"https://sso-foo.awsapps.com/start\",\
+          \"region\": \"us-west-2\",\
+          \"accessToken\": \"FakeAccessTokenContent\",\
+          \"expiresAt\": \"2022-06-03T05:53:48Z\"\
+        }\n"
+    );
+    aws_auth_library_init(allocator);
+
+    ASSERT_SUCCESS(s_sso_provider_init_config_profile(allocator, sso_test_profile_contents));
+
+    /* Redirect $HOME to the testing directory, to create the $HOME/.aws/sso/cache/<sha1>.json there. */
+    struct aws_string *actual_home = aws_string_new_from_c_str(allocator, getenv("HOME"));
+    ASSERT_SUCCESS(aws_set_environment_value(s_home_env_var, s_home_here));
+
+    /* Create access token file. */
+    struct aws_string *start_url = s_sso_extract_start_url_from_profile(allocator);
+    ASSERT_NOT_NULL(start_url);
+
+    struct aws_string *token_path = sso_access_token_path(allocator, start_url);
+    ASSERT_NOT_NULL(token_path);
+
+    ASSERT_SUCCESS(s_create_directory_components(allocator, token_path));
+    ASSERT_SUCCESS(aws_create_profile_file(token_path, sso_expired_access_token));
+
+    ASSERT_PTR_EQUALS(NULL, aws_credentials_provider_new_sso(allocator, &options));
+
+    /* Clean up */
+    ASSERT_SUCCESS(aws_set_environment_value(s_home_env_var, actual_home));
+    aws_string_destroy(actual_home);
+    aws_string_destroy(start_url);
+    aws_string_destroy(token_path);
+
+    aws_auth_library_clean_up();
+
+    return AWS_OP_SUCCESS;
+}
+AWS_TEST_CASE(
+    credentials_provider_sso_provider_access_token_expired,
+    s_credentials_provider_sso_provider_access_token_expired);
+
+static int s_credentials_provider_sso_new_destroy_from_config(struct aws_allocator *allocator, void *ctx) {
```

Comment:
Ideally, would like to see a full success path test that uses the (admittedly clumsy) function table mocks to achieve a successful credentials fetch.

Reply:
Would it be possible to do that via a follow-up PR?

Reply:
Perhaps in a future iteration?